provisioner/node_pools.go

```diff
@@ -123,13 +223,150 @@ func (p *AWSNodePoolProvisioner) Provision(ctx context.Context, nodePools []*api
 	return nil
 }
 
-func supportsT2Unlimited(instanceTypes []string) bool {
-	for _, instanceType := range instanceTypes {
-		if !strings.HasPrefix(instanceType, "t2.") {
-			return false
+func (p *KarpenterNodePoolProvisioner) kubectlExecute(ctx context.Context, args []string, stdin string) (string, error) {
+	token, err := p.tokenSource.Token()
+	if err != nil {
+		return "", err
+	}
+
+	args = append([]string{
+		"kubectl",
+		fmt.Sprintf("--server=%s", p.cluster.APIServerURL),
+		fmt.Sprintf("--token=%s", token.AccessToken),
+	}, args...)
+	if stdin != "" {
+		args = append(args, "-f", "-")
+	}
+
+	newApplyCommand := func() *exec.Cmd {
+		cmd := exec.Command(args[0], args[1:]...)
+		// prevent kubectl to find the in-cluster config
+		cmd.Env = []string{}
+		return cmd
+	}
+	var output string
+	applyManifest := func() error {
+		cmd := newApplyCommand()
+		if stdin != "" {
+			cmd.Stdin = strings.NewReader(stdin)
 		}
+		output, err = p.execManager.Run(ctx, p.logger, cmd)
+		return err
 	}
-	return true
+	err = backoff.Retry(applyManifest, backoff.WithMaxRetries(backoff.NewExponentialBackOff(), maxApplyRetries))
+	if err != nil {
+		return "", err
+	}
+	return output, nil
+}
+
+func (p *KarpenterNodePoolProvisioner) provisionNodePool(ctx context.Context, nodePool *api.NodePool, values map[string]interface{}) error {
+	template, err := p.generateNodePoolTemplate(nodePool, values, provisionersFileName)
+	if err != nil {
+		return err
+	}
+	if template == "" {
+		return nil
+	}
+	_, err = p.kubectlExecute(ctx, []string{"apply"}, template)
+	if err != nil {
+		return errors.Wrapf(err, "kubectl apply failed for node pool %s", nodePool.Name)
+	}
+	return nil
+}
+
+func (p *KarpenterNodePoolProvisioner) isKarpenterInstalled(ctx context.Context) (bool, error) {
+	gvr, err := kubernetes.ResolveKind(p.kubeMapper, crd)
+	if err != nil {
+		return false, err
+	}
+
+	_, err = p.kubeClient.Resource(gvr).Get(ctx, karpenterProvisionerResource, metav1.GetOptions{})
```

Comment:
nit: Since this is moved to the `kubernetes` package it could also be an opportunity to make the interface a bit nicer. Instead of an interface that returns multiple clients via `InitClients` it could have a client that hides the internal details of `ResolveKind` etc.

So instead of calling first `ResolveKind` and then `kubeClient.Resource(gvr).Get(ctx, karpenterProvisionerResource, metav1.GetOptions{})` you could have a method `GetCRD(ctx, resourcename, metav1.GetOptions{})` which does the `ResolveKind` internally.

This can ofc. be done later. Just came to light because of the many Exported functions in `kubernetes.go` which is hinting at a better abstraction being possible.